.github/scripts/validate_examples.py

```diff
@@ -0,0 +1,191 @@
+
+from __future__ import annotations
+
+import argparse
+import importlib.util
+import sys
+import tempfile
+import traceback
+from pathlib import Path
+from types import ModuleType
+from typing import Iterable, List, Sequence, Set, Tuple
+
+try:
+    from kfp import compiler, dsl
+except ImportError as exc:  # pragma: no cover - surfaced to callers
+    raise SystemExit(
+        "kfp is required to validate example pipelines. "
+        "Install it with `pip install kfp`."
+    ) from exc
+
+REPO_ROOT = Path(__file__).resolve().parents[2]
+DEFAULT_TARGETS = (REPO_ROOT / "components", REPO_ROOT / "pipelines")
+PIPELINE_ATTRS = ("pipeline_func", "_pipeline_func", "pipeline_spec")
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(
+        description=(
+            "Import example_pipelines.py modules for the specified components or pipelines "
+            "and compile every @dsl.pipeline function they export."
+        )
+    )
+    parser.add_argument(
+        "paths",
+        metavar="PATH",
+        nargs="*",
+        help=(
+            "Component or pipeline directories (or files within them). "
+            "If omitted, every example_pipelines.py file is validated."
+        ),
+    )
+    return parser.parse_args()
+
+
+def normalize_targets(raw_paths: Sequence[str]) -> List[Path]:
+    if not raw_paths:
+        return [target for target in DEFAULT_TARGETS if target.exists()]
+
+    normalized: List[Path] = []
+    for raw in raw_paths:
+        candidate = Path(raw)
+        if not candidate.is_absolute():
+            candidate = (REPO_ROOT / candidate).resolve()
+        else:
+            candidate = candidate.resolve()
+        if not candidate.exists():
+            raise FileNotFoundError(f"Specified path does not exist: {raw}")
+        normalized.append(candidate)
+    return normalized
+
+
+def discover_example_files(targets: Sequence[Path]) -> List[Path]:
+    discovered: List[Path] = []
+    seen: Set[Path] = set()
+
+    for target in targets:
+        search_root = target if target.is_dir() else target.parent
+
+        for candidate in _possible_example_paths(search_root):
+            if candidate in seen:
+                continue
+            try:
+                relative = candidate.relative_to(REPO_ROOT)
+            except ValueError:
+                continue
+            if relative.parts and relative.parts[0] in {"components", "pipelines"}:
+                seen.add(candidate)
+                discovered.append(candidate)
+
+    return discovered
+
+
+def _possible_example_paths(search_root: Path) -> Iterable[Path]:
+    if search_root.name == "example_pipelines.py" and search_root.is_file():
+        yield search_root
+        return
+
+    candidate = search_root / "example_pipelines.py"
+    if candidate.is_file():
+        yield candidate
+
+    for nested in search_root.rglob("example_pipelines.py"):
+        if nested.is_file():
+            yield nested
```

Comment:
could this not just be simplified into using `rglob("example_pipelines.py")` on each of the targets?

ie
```
discovered = []
for t in targets:
    for s in t.rglob("example_pipelines.py"):
        if s not in all:
            all.append(s)
```